File: 240527/트리의 부모 노드/parent-node-of-the-tree.cpp
#include <iostream>
#include <vector>

using namespace std;

int n;
vector <int> G[100000];
bool visited[100000];
int parent[100000];

void	travel(int x)
{
	for (int i = 0; i < G[x].size(); i++)
	{
		int	node = G[x][i];
		if (!visited[node])
		{
			parent[node] = x;
			visited[node] = true;
			travel(node);
		}
	}
}

int main()
{
	cin >> n;
	for (int i = 0; i < n - 1; i++)
	{
		int s, e;
		cin >> s >> e;
		G[s - 1].push_back(e - 1);
		G[e - 1].push_back(s - 1);
	}
	visited[0] = true;
	travel(0);

	for (int i = 1; i < n; i++)
		cout << parent[i] + 1 << '\n';
}

File: 240527/트리의 부모 노드/parent-node-of-the-tree.py
import sys

input = sys.stdin.readline


def travel(x, G, visited, parent):
    for i in G[x]:
        if not visited[i]:
            visited[i] = True
            parent[i] = x
            travel(i, G, visited, parent)

n = int(input())

G = [[] for _ in range(n)]
visited = [False for _ in range(n)]
parent = [0 for _ in range(n)]

for _ in range(n - 1):
    s, e = map(int, input().split())
    G[s - 1].append(e - 1)
    G[e - 1].append(s - 1)

travel(0, G, visited, parent)

for i in parent[1:]:
    print(i + 1)